feat: Readcache buffers pool

dmga44 opened the pull request:
This PR targets reducing the allocation and deallocation of memory
when actively reading. Should improve performance on heavy read
workloads.

Changes:
- implement `ReadCacheEntriesPool` to hold the read cache entries for a
while (1s by default, configurable via parameters) after supposedly
being deleted from cache.
- when requesting a new entry in the ReadCache we'd look at the global
`ReadCacheEntriesPool` singleton and take one from there if available.
- a new thread for periodically cleaning the `ReadCacheEntriesPool` is
created, it checks every 0.1 seconds.
- the operations on the `ReadCacheEntriesPool` are thread safe. A mutex
ensures protection.

As side changes:
- most of the readdata_cache.h function implementations were moved to
readdata_cache.cc altogether with the new `ReadCacheEntriesPool`
functions.
- various clang tidy warnings were addressed like removing unused
includes.

The client read algorithm params for LongSystemTests suite has been
decreased to save some memory.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat: Readcache buffers pool

Changes:
- implement ReadCacheEntriesPool to hold the read cache entries for a
while (10s) after supposedly being deleted from cache.
- when requesting a new entry in the ReadCache we'd look at the global
ReadCacheEntriesPool singleton and take one from there if available.
- a new thread for periodically cleaning the ReadCacheEntriesPool is
created, it checks every 1 second.
- the operations on the ReadCacheEntriesPool are thread safe. A mutex
ensures protection.

As side changes:
- most of the readdata_cache.h function implementations were moved to
readdata_cache.cc altogether with the new ReadCacheEntriesPool
functions.
- varios clang tidy warnings were addressed like removing unused
includes.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 commented on src/mount/readdata_cache.h:
```diff
@@ -338,191 +204,73 @@ class ReadCache {
-		}
+class ReadCacheEntriesPool {
+public:
+	ReadCacheEntriesPool() : entriesMap_(), mutex_() {
```
These initializers are redundant, please consider to remove the explicit initialization.

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,446 @@
+		// Delete buffers outside the lock to avoid contention overhead.
+		for (auto &entry : entriesToDelete) { delete entry; }
+	}
+}
```
Add empty line at the end of the file.

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,446 @@
+}
+
+void ReadCacheEntriesPool::cleanerThreadFunc_(std::stop_token &stopToken) {
+	pthread_setname_np(pthread_self(), "ReadCacheEntriesPool's cleanerThread");
```
I think thread names can be up to 16 characters to be shown, please check it.

dmga44 replied:
Changed to "RCEP cleaner".

lgsilva3087 commented on src/mount/readdata_cache.h:
```diff
@@ -338,191 +204,73 @@ class ReadCache {
-		while (it != entries_.end()) {
-			it = erase(it);
-		}
+class ReadCacheEntriesPool {
```
As this class manage its own resources and looks like is intended to be used as a singleton. Consider to remove some non-expected constructors/operators to explicitely avoid misuse, like:

```cpp
ReadCacheEntriesPool(const ReadCacheEntriesPool&) = delete;
ReadCacheEntriesPool& operator=(const ReadCacheEntriesPool&) = delete;
ReadCacheEntriesPool(ReadCacheEntriesPool&&) = delete;
ReadCacheEntriesPool& operator=(ReadCacheEntriesPool&&) = delete;
```

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/readdata_cache.h:
```diff
@@ -338,191 +204,73 @@ class ReadCache {
-		return ret;
-	}
+		BufferEntry(BufferEntry &&other) noexcept
+		    : entry_(other.entry_), timer_(other.timer_) {}
```
Consider:
```cpp
BufferEntry(BufferEntry &&other) noexcept
    : entry_(other.entry_), timer_(std::move(other.timer_)) {
    other.entry_ = nullptr;
}
```
The moved object should be put into a valid but unspecified stated after move. As entry_ is a raw pointer, the common practice is to set it to nullptr.

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/readdata_cache.h:
```diff
@@ -338,191 +204,73 @@ class ReadCache {
-			--it;
-		}
+private:
+	void cleanerThreadFunc_(std::stop_token &stopToken);
```
Consider to pass the stop token by value, which matches how std::jthread passes the token from the lambda to the thread function.
`void cleanerThreadFunc_(std::stop_token stopToken);`

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,446 @@
+
+ReadCache::EntrySet::iterator ReadCache::insert(EntrySet::iterator it, Offset offset, Size size) {
+	it = clearCollisions(it, offset + size);
+	// Entry *e = new Entry(offset, size);
```
Remove the comment for the final version.

dmga44 replied:
Removed.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,446 @@
+		std::unique_lock usedMemoryLock(gReadCacheMemoryMutex);
+		decreaseUsedReadCacheMemory(e->buffer.size());
+		usedMemoryLock.unlock();
+		// delete e;
```
Remove the comment for the final version.

dmga44 replied:
Done.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,446 @@
+	buffers.pop_back();
+	lock.unlock();
+
+	// Reinitialize the entry
+	entry->offset = offset;
+	entry->buffer.clear();
+	entry->timer = Timer();
+	entry->refcount = 0;
+	entry->requested_size = size;
+	entry->done = false;
```
Consider a function, member of ReadCache::Entry to reset the values, taking as arguments offset and size. Something like:
`entry->resetValues(offset, size);`
Consider it only if this will be used in other places.

dmga44 replied:
It is not used anywhere else, so I did not add it.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,444 @@
+void ReadCache::selective_clear(uint32_t chunkIndex) {
+	// The least starting offset such that the end of the entry could overlap with the chunk
+	uint64_t leastOverlappingOffset = 0;
+	if (chunkIndex > 0) { leastOverlappingOffset = (chunkIndex - 1) * 1024 * 1024 * 64; }
+	uint64_t chunkStart = chunkIndex * 1024 * 1024 * 64;
+	uint64_t chunkEnd = chunkStart + 1024 * 1024 * 64;
```
Please check the comment from Gemini

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/readdata_cache.cc:
```diff
@@ -0,0 +1,446 @@
+		decreaseUsedReadCacheMemory(e->buffer.size());
+		usedMemoryLock.unlock();
+		// delete e;
+		gReadCacheEntriesPool->putEntry(e);
```
Consider to check gReadCacheMemoryAlmostExceeded before this call. If there is too much pressure in the memory, maybe we can just delete the entry.

dmga44 replied:
The `putEntry` call was substituted by the following in `erase` and `clearReserved` functions:
```cpp
// If we are almost out of memory, delete the entry instead of putting it back to the pool
if (gReadCacheMemoryAlmostExceeded) {
	delete e;
} else {
	gReadCacheEntriesPool->putEntry(e);
}
```

dmga44 replied:
As you may notice, the memory held in the `ReadCacheEntriesPool` is not being computed in the `gUsedReadCacheMemory`, this is mainly because of two reasons:
- it is actually not in cache, as it does not hold any valuable data, just the entries.
- it would be a complete mess to keep track of it (mainly because the resizes of the buffers in the read algorithm won't be actually allocating data)

lgsilva3087 reviewed (commented):
Please see my comments.

dmga44 commented on src/mount/readdata.cc:
```diff
@@ -607,6 +600,7 @@ void read_data_init(uint32_t retries,
 
 	readDataTerminate = false;
 
```
Applied.

dmga44 committed:
feat: Add readbuffersexpirationtime option

The living time of the entries inside the ReadCacheEntriesPool should
be configurable to avoid consuming too much memory. The proposed
default value is 1 second but the recycle pattern of this structure
should make most entries don't stay that long.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 reviewed (approved):
LGTM

dmga44 committed:
test: Reduce read params for Longs

Signed-off-by: Dave <dave@leil.io>
